esp/esp/program/modules/handlers/studentregphasezero.py

```diff
@@ -60,6 +64,69 @@ def module_properties(cls):
 
     def lottery(self, prog, role):
         #run lottery
+
+        def RepresentInt(x):
+            try:
+                int(x)
+                return True
+            except ValueError:
+                return False
+
+        sibgroups = {}
+        indiv = {}
+
+        # Get grade caps and process student data in PhaseZeroRecords
+        grade_caps_tag = Tag.getProgramTag('program_size_by_grade', prog)
+        grade_caps_tag_split = [x for x in re.split('(\d+)', grade_caps_tag) if RepresentInt(x)]
+        grade_caps = {grade_caps_tag_split[2*i]:int(grade_caps_tag_split[2*i+1]) for i in range(len(grade_caps_tag_split)/2)}
+
+        # Can you check if these are the correct calls to methods/parameters?
+        for entry in PhaseZeroRecord.objects:
+            group_number = entry.lottery_number
+            username = entry.user
+            user_grade = entry.user.getGrade(prog)
+
+            indiv[username] = (group_number, user_grade)
+            if group_number not in sibgroups.keys():
+                sibgroups[group_number] = [(username, user_grade)]
+            else:
+                sibgroups[group_number].append((username, user_grade))
+
+        ###############################################################################
+        # The lottery algorithm is run, with randomization and processing in order.
+        # If any one in the group doesn't get in (due to cap size), no one in that group gets in.
+
+        # not sure if we need to deep copy here; kept as shallow
+        groups = sibgroups.keys()
+        order = copy.copy(groups)
+        random.shuffle(order)
+
+        counts = {key:0 for key in grade_caps}
+        winners = Group(name=(str(prog) + ' Winners'))
+        waitlist = Group(name=(str(prog) + ' Waitlist'))
```

Comment:
Hmm, since we don't plan to add any permissions to the waitlist group, I don't think this is necessary.